Feature/m0018 pqt 144 run test suite (#87)
MicroEJ/docs: Tutorials/tutorialRunATestSuite.rst

  to learn more about creating custom Foundation Library.

The Test Suite Engine
---------------------

The Test Suite Engine (TSE) is the process responsible for building,
running a Test Suite and providing a report.

A Test Suite contains one or more JUnit tests.  For each test, the Test Suite Engine will:

1. Create MicroEJ Application that calls the test.

2. Build a MicroEJ Firmware from the MicroEJ Application.

3. Flash MicroEJ Firmware onto the device.

4. Collect the result of the execution.

5. Append the result to the Test Report.

6. Repeat until all tests of the Test Suite have been executed.

.. figure:: images/test-suite-engine-overview.PNG
   :align: center

Configuring the Test Suite
--------------------------

Configuring the BSP Connection of the Test Suite
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Several properties must be defined depending on the type of BSP
Connection used by the MicroEJ Platform.

For a MicroEJ Application, this properties are set using the launcher
of the application.  For a Test Suite, the properties are defined in a
file name ``config.properties`` in the root folder of the Test Suite.
For example,

> https://github.com/MicroEJ/PlatformQualificationTools/tree/master/tests/fs/java/java-testsuite-fs/config.properties.

Author: fixed

``run.sh`` script.

If that is not the case, the Serial To Socket Transmitter MicroEJ Tool
should be used.  This tool can be configured to listen on a particular
COM port and redirect the output on a local socket.  The properties
``microej.testsuite.properties.testsuite.trace.ip`` and
``microej.testsuite.properties.testsuite.trace.port`` must be
configured.

Follow the following steps to create a launcher for Serial To Socket
Transmitter:

- :guilabel:`Run` > :guilabel:`Run Configurations...`.
- Right-click on :guilabel:`MicroEJ Tool` > :guilabel:`New`.
- In the :guilabel:`Execution` tab:
  - Set :guilabel:`Name` to ``Serial To Socket Transmitter``.
  - Select a MicroEJ Platform available in the workspace in :guilabel:`Target` > :guilabel:`Platform`.
  - Select ``Serial To Socket Transmitter`` in :guilabel:`Execution` > :guilabel:`Settings`.
  - Set the :guilabel:`Output folder` to the workspace.
- In the :guilabel:`Configuration` tab:
  - Set the correct COM port and baudrate for the device in :guilabel:`Serial Options`.
  - Set a valid port number in :guilabel:`Server Options` >
    :guilabel:`Port`.  This port is the same as the one set in
    ``config.properties`` as

>     ``microej.testsuite.properties.testsuite.trace.port``.

Reviewer: Rendering of this section is not good in github preview. I don't know if the problem comes from github or the RST.

Author: fixed

Configuring the Test Suite Specifics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Depending on the Test Suite and the specificities of the device,
various properties may be required and adjusted.  See the file
``validation/microej-testsuite-common.properties`` (for example

> https://github.com/MicroEJ/PlatformQualificationTools/tree/master/tests/fs/java/java-testsuite-fs/validation/microej-testsuite-common.properties)

Reviewer: We should not reference a specific file in the master branch.
Prefer referencing a commit or a tag.

Author: fixed.

and the README of the Test Suite for a description of each property.

Running the Test Suite
----------------------

> To run the Test Suite, right clic on the Test Suite module and select
> ``Build Module``.

Reviewer: ```suggestion
To run the Test Suite, right click on the Test Suite module and select
``Build Module``.
```

Skipping Tests
--------------

It is possible to exclude some tests from being executed by the Test
Suite Engine.

In the following example, only the classes that match
``TestFilePermission`` are executed.

.. code:: properties

          # Comma separated list of patterns of files that must be included	
          # test.run.includes.pattern=**/Test*.class
          test.run.includes.pattern=**/TestFilePermission*.class
          # Comma separated list of patterns of files that must be excluded (defaults to inner classes)
          test.run.excludes.pattern=**/*$*.class

> Several reasons might explain why exclude some tests:

Reviewer: ```suggestion
Several reasons might explain why to exclude some tests:
```

> - Iterative development: Test only the ADLA that is currently being

Reviewer: ```suggestion
- Iterative development: Test only the abstraction layer that is currently being
```

  developed.  The full Test Suite must still be ran to validate the
  complete implementation.

- Known bugs in the Foundation Library: The latest version of the Test
  Suite for a given Foundation Library might contains regression
  tests, or tests for new features.  If the MicroEJ Platform doesn't

>   use the latest Foundation Library, then it is necessary to exclude

Reviewer: ```suggestion
  use the latest Foundation Library, then it can be necessary to exclude
```

  the new tests.

- Known bugs in the Foundation Library implementation: The project

>   might have specific requirements that prevents a fully compliant

Reviewer: ```suggestion
  might have specific requirements that prevent a fully compliant
```

  implementation of the Foundation Library.

The Test Suite Report
---------------------

Once the Test Suite is completed, the Test Suite Report is stored in
``target~/test/html/junit-noframes.html``.

At the beginning of the file, a summary is displayed.  Below, all
execution traces for each test executed are available.

> If necessary, the binary produced and ran on the device by the Test

Reviewer: ```suggestion
If necessary, the binaries produced and ran on the device by the Test
```

Suite Engine are available in

> ``target~/test/xml/TIMESTAMP/bin/FULLY-QUALIFIED-CLASSNAME/application.out``.

Reviewer: ```suggestion
``target~/test/xml/<TIMESTAMP>/bin/<FULLY-QUALIFIED-CLASSNAME>/application.out``.
```

> The following image shows a example of a Test Suite Report for the Test Suite FS.

Reviewer: ```suggestion
The following image shows an example of a Test Suite Report for the Test Suite FS.
```

Understand How to Run a Test Suite for a MicroEJ Platform
=========================================================

This tutorial explains how to run a Test Suite on a MicroEJ Platform
using the `Platform Qualification Tools
<https://github.com/MicroEJ/PlatformQualificationTools>`_.  The
file system Test Suite will be used as an example.

This tutorial assumes the following:

- Good knowledge of the :ref:`chapter-glossary`.

- :ref:`tutorialUnderstandMicroEJFirmwareBuild`.

> - A MicroEJ Platform with:

Reviewer 2: A MicroEJ Platform source imported in the MicroEJ SDK workspace:

Author: fixed

  - An implementation of :ref:`LLFS-API-SECTION` version 1.0.2 in
    ``com.microej.pack#fs-4.0.2``.

  - A partial or full :ref:`bsp_connection`.

- MicroEJ SDK distribution 20.07 (see :ref:`get_sdk_version`)

This tutorial does not explain how to create a Test Suite.  See
:ref:`testsuite` for it.

Foundation Library Test Suite
-----------------------------

The purpose of a MicroEJ Foundation Library Test Suite is to validate
the implementation of the Low Level API specified by the Foundation
Library.

In essence, a Foundation Library provides an API to be used by a
MicroEJ Application or an Add-on Library.

.. figure:: ../overview/images/ClassPath_2.png
   :alt: MicroEJ Foundation Libraries and Add-On Libraries
   :align: center

   MicroEJ Foundation Libraries, Add-On Libraries and MicroEJ Application

For example, the Java file system API ``java.io`` is provided by the

> MicroEJ Foundation Library named FS.  It is the role of the MicroEJ

Reviewer 2: MicroEJ Platform Engineer: wtf is this and where is it defined?
Prefer: The Abstraction layer of each foundation API must be implemented in the BSP.

Author: fixed

> Platform Engineer to implement the required functions, called

Reviewer 2: Avoid platform engineer (never defined before, too complex)

Author: fixed

> Abstraction Layer, in the BSP.  The Test Suite is used to validate the

Reviewer 2: First time BSP is mentionned. What is BSP?

Author: The tutorial assumes the reader has read the tutorial on how to build a MicroEJ Firmware which explains what is the BSP. Do we need to repeat it?

Reviewer 2: ok noted.

> implementation.

Reviewer 2: to validate the C code implementation of the Abstraction Layer.

Author: fixed

.. note::

  See `How to Setup a New Foundation Library
  <https://github.com/MicroEJ/How-To/tree/1.8.3/FoundationLibrary-Get-Started>`_
  to learn more about creating custom Foundation Library.

The Test Suite Engine
---------------------

The Test Suite Engine (TSE) is the process responsible for building,
running a Test Suite, and providing a report.

A Test Suite contains one or more JUnit tests.  For each test, the Test Suite Engine will:

1. Create MicroEJ Application that calls the test.

2. Build a MicroEJ Firmware from the MicroEJ Application.

> 3. Flash MicroEJ Firmware onto the device.

Reviewer 2: the MicroEJ Firmware...

Author: fixed

4. Collect the result of the execution.

5. Append the result to the Test Report.

6. Repeat until all tests of the Test Suite have been executed.

.. figure:: images/test-suite-engine-overview.PNG
   :align: center

Configuring the Test Suite
--------------------------

Configuring the BSP Connection of the Test Suite
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Several properties must be defined depending on the type of BSP
Connection used by the MicroEJ Platform.

For a MicroEJ Application, these properties are set using the launcher
of the application.  For a Test Suite, the properties are defined in a
file named ``config.properties`` in the root folder of the Test Suite.
For example,
https://github.com/MicroEJ/PlatformQualificationTools/blob/2.3.0/tests/fs/java/java-testsuite-fs/config.properties.

See :ref:`bsp_connection` for an explanation of the properties.  See
the comments in the file for a details description of each properties.
The ``microej.testsuite.properties.deploy.*`` and
``target.platform.dir`` properties are required.

Collecting the Results of Execution
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When the Test Suite is executed, the Test Suite Engine must read the

Author: I checked with adiallo and he confirmed that the default behavior for the Test Suite Engine as configured in the PQT is to expect the `run.bat` to output on stdout the trace of execution. If the user sets the address and port then it uses it.

> When the Test Suite is executed, the Test Suite Engine must read the

Author: Removed the mention of redirection to stdout.
